### PR TITLE
Add media version history list

### LINKED ISSUE #931: Add media version history list to overlay

### PR DESCRIPTION
**Pull request by alexander-schranz**
Add Version History to media edit overlay:

API Object changed BC BREAK!

__Tasks:__

- [x] test coverage
- [ ] gather feedback for my changes
- [x] submit changes to the documentation
- [x] added changelog line

__Informations:__

| Q             | A
| ------------- | ---
| Tests pass?   | yes
| Fixed tickets | fixes #931 
| BC Breaks     | yes
| Doc           | https://github.com/sulu-cmf/docs/pull/14

bildschirmfoto 2015-03-20 um 14 48 57

### CODE REVIEW

File: src/Sulu/Bundle/MediaBundle/Api/Media.php [hunk 1]
```diff
@@ -249,13 +278,17 @@ public function getVersion()
...
     {
         $versions = array();
-        /** @var File $file */
-        foreach ($this->entity->getFiles() as $file) {
-            /** @var FileVersion $fileVersion */
-            foreach ($file->getFileVersions() as $fileVersion) {
-                $versions[] = $fileVersion->getVersion();
+        /** @var FileVersion $fileVersion */
+        foreach ($this->getFile()->getFileVersions() as $fileVersion) {
+            $versionData = array();;
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Api/Media.php]: 2 `;`

File: src/Sulu/Bundle/MediaBundle/Api/Media.php [hunk 2]
```diff
@@ -653,6 +686,25 @@ private function getFileVersion()
...
+            return $this->file;
+        }
+
+        /** @var File $file */
+        foreach ($this->entity->getFiles() as $file) {
+            // currently only one file per media exists
+            $this->file = $file;
+            return $this->file;
+        }
+        throw new FileNotFoundException($this->entity->getId(), $this->version);
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Api/Media.php]: Would put a new line before this one

File: src/Sulu/Bundle/MediaBundle/Media/Exception/FileNotFoundException.php [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ * @package Sulu\Bundle\MediaBundle\Media\Exception
+ */
+class FileNotFoundException extends MediaException
+{
+    /**
+     * @param int $id
+     */
+    public function __construct($id)
+    {
+        parent::__construct('File from the Media with ID ' . $id . ' not found', self::EXCEPTION_CODE_FILE_NOT_FOUND);
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Media/Exception/FileNotFoundException.php]: Put the id in quotes.

File: src/Sulu/Bundle/MediaBundle/Resources/translations/sulu/backend.de.xlf [hunk 4]
```diff
@@ -114,6 +114,18 @@
                 <source>sulu.media.edit.loading</source>
                 <target>Medien werden geladen</target>
             </trans-unit>
+            <trans-unit id="243d88cf2f64652633a5ca2169f22258" resname="sulu.media.history">
+                <source>sulu.media.history</source>
+                <target>History</target>
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/translations/sulu/backend.de.xlf]: @alexander-schranz Do we also use `History` as the german translation? /cc @berndhep

**Comment by danrot** [comment]
@alexander-schranz If you have a BC break you should put some lines for explanation into the `UPGRADE.md` file

File: src/Sulu/Bundle/MediaBundle/Media/Exception/FileNotFoundException.php [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+ * (c) MASSIVE ART WebServices GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\MediaBundle\Media\Exception;
+
+/**
+ * @package Sulu\Bundle\MediaBundle\Media\Exception
```
Comment (dantleech) [inline src/Sulu/Bundle/MediaBundle/Media/Exception/FileNotFoundException.php]: Remove `@package`

File: src/Sulu/Bundle/MediaBundle/Media/Exception/FileNotFoundException.php (hunk 3 above)
Comment (dantleech) [inline src/Sulu/Bundle/MediaBundle/Media/Exception/FileNotFoundException.php]: Also use the `sprintf` syntax:

````
__Construct(sprintf('File from the Media with ID "%s" not found', $id));
````

File: src/Sulu/Bundle/MediaBundle/Media/Exception/MediaException.php [hunk 6]
```diff
@@ -138,6 +138,12 @@ class MediaException extends Exception
      */
     const EXCEPTION_FILENAME_ALREADY_EXISTS = 5020;
 
+    /**
+     * @var int
+     * @description this exception code is thrown when the file is not found
```
Comment (dantleech) [inline src/Sulu/Bundle/MediaBundle/Media/Exception/MediaException.php]: `@description` is redundant, and not part of phpdoc syntax: http://www.phpdoc.org/docs/latest/index.html

````
/**
 * This exception code is thrown when the file is not found
 * @var int
 */
````

File: src/Sulu/Bundle/MediaBundle/Media/Exception/FileNotFoundException.php (hunk 5 above)
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Media/Exception/FileNotFoundException.php]: And add some real description

File: src/Sulu/Bundle/MediaBundle/Media/Exception/MediaException.php [hunk 7]
```diff
@@ -138,6 +154,13 @@ class MediaException extends Exception
      */
     const EXCEPTION_FILENAME_ALREADY_EXISTS = 5020;
 
+    /**
+     * this exception code is thrown when the file is not found
```
Comment (dantleech) [inline src/Sulu/Bundle/MediaBundle/Media/Exception/MediaException.php]: Sentences should always start with a capital letter

File: src/Sulu/Bundle/MediaBundle/Media/Manager/DefaultMediaManager.php [hunk 8]
```diff
@@ -822,9 +826,32 @@ public function addFormatsAndUrl(Media $media)
...
+
+        $media->setAdditionalVersionData($versionData);
+
+        // Set Current Url
+        if (
+            isset($versionUrls[$media->getVersion()])
+            && isset($versionUrls[$media->getVersion()]['url'])
+        ) {
+            $media->setUrl($versionUrls[$media->getVersion()]['url']);
+        } else {
```
Comment (dantleech) [inline src/Sulu/Bundle/MediaBundle/Media/Manager/DefaultMediaManager.php]: `else` is not required here (it should [generally be avoided](http://williamdurand.fr/2013/06/03/object-calisthenics/#2-don-t-use-the-else-keyword)) as you can return `$media` on the previous line

File: src/Sulu/Bundle/MediaBundle/Tests/Functional/Controller/MediaControllerTest.php [hunk 9]
```diff
@@ -746,6 +748,7 @@ public function testFileVersionUpdateWithoutDetails()
         $this->assertEquals($this->mediaDefaultDescription, $response->description);
         $this->assertEquals($this->collection->getId(), $response->collection);
         $this->assertEquals(2, $response->version);
+        $this->assertEquals(2, count((array)$response->versions));
```
Comment (dantleech) [inline src/Sulu/Bundle/MediaBundle/Tests/Functional/Controller/MediaControllerTest.php]: `->assertCount(2, (array) $response->versions)`

File: UPGRADE.md [hunk 10]
```diff
@@ -2,6 +2,10 @@
 
 ## dev-develop
 
+### Media API Object
+
+The media API Object the versions attribute changed from [array to object list](https://github.com/sulu-io/docs/pull/14/files).
```
Comment (danrot) [inline UPGRADE.md]: Would write:

```
The `versions` attribute of the media API object changed ...
```

**Comment by wachterjohannes** [comment]
can i download a specific version of an image?

**Comment by alexander-schranz** [comment]
@wachterjohannes that what the history is for

**Pull request merged**